ultralytics/utils/metrics.py

```diff
@@ -64,6 +64,9 @@ def box_iou(box1, box2, eps=1e-7):
         (torch.Tensor): An NxM tensor containing the pairwise IoU values for every element in box1 and box2.
     """
 
+    # NOTE: need float32 to get accurate iou values
+    box1 = torch.as_tensor(box1, dtype=torch.float32)
+    box2 = torch.as_tensor(box2, dtype=torch.float32)
```

Comment:
@Laughing-q this function is extremely optimized, I don't think making a copy of the variables is the fastest way, maybe we can inline this into the box.unsqueeze line directly?

```diff
- (a1, a2), (b1, b2) = box1.unsqueeze(1).chunk(2, 2), box2.unsqueeze(0).chunk(2, 2)
+ (a1, a2), (b1, b2) = box1.float().unsqueeze(1).chunk(2, 2), box2.float().unsqueeze(0).chunk(2, 2)
```

Reply:
@glenn-jocher I thought `as_tensor` is the one api that reuse the tensor(without any copy) if it already has the requested dtype and `.float()` is the one would copy tensor anyway whether if it has requested dtype. But I just revisited pytorch docs and found `.float()` behaviors just like `as_tensor` for dtype converting.
https://pytorch.org/docs/2.3/generated/torch.as_tensor.html#torch-as-tensor
https://pytorch.org/docs/2.3/generated/torch.Tensor.to.html#torch.Tensor.to
So yes I guess we can directly inline it into one line with box.unsqueeze.

Reply:
updated https://github.com/ultralytics/ultralytics/pull/13113/commits/3ecfbb92e92ba46b9888d76681af5100380c014d